onedal/ensemble/forest.cpp

```diff
@@ -0,0 +1,294 @@
+/*******************************************************************************
+* Copyright 2023 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/decision_forest.hpp"
+
+#include "onedal/common.hpp"
+
+namespace py = pybind11;
+
+namespace oneapi::dal::python {
+
+template <typename Task, typename Ops>
+struct method2t {
+    method2t(const Task& task, const Ops& ops) : ops(ops) {}
+
+    template <typename Float>
+    auto operator()(const py::dict& params) {
+        using namespace decision_forest;
+
+        const auto method = params["method"].cast<std::string>();
+
+        ONEDAL_PARAM_DISPATCH_VALUE(method, "dense", ops, Float, method::dense);
+        ONEDAL_PARAM_DISPATCH_VALUE(method, "hist", ops, Float, method::hist);
+        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
+    }
+
+    Ops ops;
+};
+
+std::vector<std::string> split(const std::string& str) {
+    const size_t size = str.size();
+    std::vector<std::string> result;
+
+    size_t start = 0;
+    for (size_t i = 0; i < size; ++i) {
+        if (str[i] == '|') {
+            result.emplace_back(str.substr(start, i - start));
+            start = i + 1;
+        }
+        else if (i == size - 1) {
+            result.emplace_back(str.substr(start, i - start + 1));
+        }
+    }
+
+    return result;
+}
+
+auto get_error_metric_mode(const py::dict& params) {
+    using namespace decision_forest;
+
+    auto mode = params["error_metric_mode"].cast<std::string>();
+    auto modes = split(mode);
+    const size_t modes_num = modes.size();
+
+    error_metric_mode result_mode = error_metric_mode::none;
+    for (size_t i = 0; i < modes_num; ++i) {
+        if (modes[i] == "none")
+            result_mode |= error_metric_mode::none;
+        else if (modes[i] == "out_of_bag_error")
+            result_mode |= error_metric_mode::out_of_bag_error;
+        else if (modes[i] == "out_of_bag_error_per_observation")
+            result_mode |= error_metric_mode::out_of_bag_error_per_observation;
+        else
+            ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(mode);
+    }
+    return result_mode;
+}
```

Comment:
It seems this is a common task: to convert a string with parameters like `mode` into a combination of enum values, i.e. `result_mode`.
Let's use some unified approach for solving this task.
neighbors.cpp (get_onedal_result_options function) uses regular expressions for handling the input string. Can you re-write get_error_metric_mode and get_infer_mode in the same manner?
You won't need a separate `split` function in this case.